DevWs for Most Target Limited containing containing Zigbee Multi Button

adurosmart opened the pull request.

adurosmart commented:
@greens Help me to review this code,we add code of dimming button and scene button ,Thank you very much.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -30,7 +30,12 @@ metadata {
+        fingerprint inClusters: "0000, 0003, 0004, 0005, 0006, 0008, 0300, FCCC, 1000", outClusters: "0000, 0003, 0004, 0005, 0006, 0008, 0300, FCCC, 1000", manufacturer: "AduroSmart Eria", model: "ADUROLIGHT_CSC", deviceJoinName: "Eria scene button switch V2.1"
+        fingerprint inClusters: "0000, 0003, 0004, 0005, 0006, 0008, 0300, FCCC, 1000", outClusters: "0000, 0003, 0004, 0005, 0006, 0008, 0300, FCCC, 1000", manufacturer: "ADUROLIGHT", model: "ADUROLIGHT_CSC", deviceJoinName: "Eria scene button switch V2.0"
+        fingerprint inClusters: "0000, 0003, 0008, FCCC, 1000", outClusters: "0003, 0004, 0006, 0008, FCCC, 1000", manufacturer: "AduroSmart Eria", model: "Adurolight_NCC", deviceJoinName: "Eria dimming button switch V2.1"
+        fingerprint inClusters: "0000, 0003, 0008, FCCC, 1000", outClusters: "0003, 0004, 0006, 0008, FCCC, 1000", manufacturer: "ADUROLIGHT", model: "Adurolight_NCC", deviceJoinName: "Eria dimming button switch V2.0"
```
you'll need to add the 

`, mnmn: "SmartThings", vid: "generic-4-button"`

from the above lines to have this display correctly.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -56,6 +61,12 @@ def parse(String description) {
 	if (result.name == "switch") {
 		result = createEvent(descriptionText: "Wake up event came in", isStateChange: true)
 	}
+    if ((device.getDataValue("model") == "Adurolight_NCC") || (device.getDataValue("model") == "ADUROLIGHT_CSC")) {
```
it would be better to add this inside `parseAttrMessage` to maintain the parallel structure.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -203,17 +217,34 @@ private getButtonMap() {[
 
 private getSupportedButtonValues() {
 	def values = ["pushed", "held"]
+    if ((device.getDataValue("model") == "Adurolight_NCC") || (device.getDataValue("model") == "ADUROLIGHT_CSC")) {
```
you've used this same conditional in several places. It's probably best for readability to break it out into its own function. Perhaps `isAduroSmartRemote()`?

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +254,77 @@ private getModelBindings(model) {
+private Map parseAduroSmartButtonMessage(Map descMap){
+    def buttonState = ""
+    def buttonNumber = 0
+    if (descMap.clusterInt == 0x0006) {
```
Please use constants instead of magic numbers for clusters.

0x0006 is `zigbee.ONOFF_CLUSTER`
0x0008 is `zigbee.LEVEL_CONTROL_CLUSTER`

for 0xFCCC define a function 

`def getMANUFACTURER_SPECIFIC_CLUSTER() {0xFCCC}`

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -203,17 +217,34 @@ private getButtonMap() {[
+        "Adurolight_NCC" : 4
 ]}
 
 private getModelBindings(model) {
```
I'm worried that since your devices report using the level control cluster and you're not setting up any bindings here that they may not actually report when they are pressed.

greens replied:
same for 0xFCCC

tpmanley replied:
@adurosmart did you take a look at this comment? I think you at least need the level control binding and likely the 0xFCCC binding unless you handle that cluster automatically without a binding.

adurosmart replied:
@tpmanley  Thank you review, i add LEVEL_CONTROL_CLUSTER  bindings.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +254,77 @@ private getModelBindings(model) {
 }
+
+private Map parseAduroSmartButtonMessage(Map descMap){
+    def buttonState = ""
```
this value will never be anything except "pushed"

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +254,77 @@ private getModelBindings(model) {
+        //log.debug "descMap.data is $descMap.data"
+        def list2 = descMap.data
+        //log.debug "list2 is $list2"
+        def keynumber = list2[1]
```
`buttonNumber = (list2[1] as int) + 1`

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +254,77 @@ private getModelBindings(model) {
+        def descriptionText = "$device.displayName button $buttonNumber was $buttonState"
+        //log.debug "descriptionText is $descriptionText"
+        state.buttonNumber = buttonNumber
+        state.pressTime = now()
```
these lines here and above are only used when the device sends a different message when the button is pressed down and when it is released so that we can emulate "pushed" and "held". You are only doing "pushed" here, so this is unnecessary.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +254,77 @@ private getModelBindings(model) {
+            state.pressTime = now()
+            def childevent = createEvent(name: "button", value: "pushed",data: [buttonNumber: 1], isStateChange: true)
+            sendEventToChild(buttonNumber, childevent)
+            return createEvent(name: "button", value: buttonState, data: [buttonNumber: buttonNumber], descriptionText: descriptionText, isStateChange: true)
```
you duplicate these exact 3 lines of code in 3 separate if statements instead of just putting it after them

greens reviewed (changes requested):
On top of specific requests, please use tabs instead of spaces to match the existing file.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -203,17 +215,34 @@ private getButtonMap() {[
-		"3450-L2" : 4
+		"3450-L2" : 4, 
+        "ADUROLIGHT_CSC" : 4,
+        "Adurolight_NCC" : 4
```
using spaces instead of tabs

adurosmart replied:
Fixed.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +252,45 @@ private getModelBindings(model) {
+			buttonNumber = 3
+		    }
+		}
+	}else if (descMap.clusterInt == getAduroSmart_SPECIFIC_CLUSTER()) {
```
thanks to some nice syntactic sugar in groovy, you can just type `AduroSmart_SPECIFIC_CLUSTER` without the `get` or `()`. It's essentially the Groovy way to do constants, which is why the whole function name (except for `get`) is typically capitalized.

greens replied:
see: `getButtonMap()`

adurosmart replied:
fixed.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -174,6 +183,9 @@ private addChildButtons(numberOfButtons) {
 			String childDni = "${device.deviceNetworkId}:$endpoint"
 			def componentLabel = (device.displayName.endsWith(' 1') ? device.displayName[0..-2] : device.displayName) + "${endpoint}"
+			if (isAduroSmartRemote()) {
+				componentLabel = device.displayName + " child button ${endpoint}"
```
just so you know, your display names will be absurdly long and most likely be truncated, making it difficult to determine which button is which. For example:

`Eria scene button switch V2.0 child button 1`

Perhaps consider changing your join names to `Eria Remote` or something like that. If you like it how it is, that's fine, I'm just warning you about usability concerns I have.

adurosmart replied:
i using " - " instead of "child button"

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -63,6 +68,10 @@ def parse(String description) {
+	if (isAduroSmartRemote()) {
+		def descMap = zigbee.parseDescriptionAsMap(description)
+		result = parseAduroSmartButtonMessage(descMap)
+    	}
```
Reconsider the placement of this `if` statement. I'm worried it might result in errors when the device reports its battery level.

greens replied:
And the results will be overwritten when the device sends ONOFF messages.

It might fit as an `else if` after the power config cluster statement

adurosmart replied:
@greens 
Thanks.
Fixed.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -203,17 +215,34 @@ private getButtonMap() {[
+			"01" : 1,
+			"02" : 2,
+			"03" : 3,
+			"04" : 4
```
the code using `getButtonMap` is not intended to be used in the flow you've outlined I'm pretty sure, making these additions unnecessary.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +240,45 @@ private getModelBindings(model) {
+		sendEventToChild(buttonNumber, childevent)
+		def descriptionText = "$device.displayName button $buttonNumber was $buttonState"
+		return createEvent(name: "button", value: buttonState, data: [buttonNumber: buttonNumber], descriptionText: descriptionText, isStateChange: true)
+        }else {
```
nit: spaces instead of tabs again

your editor should have a function that allows you to convert spaces->tabs all at once

greens commented:
@tpmanley before I sign off on this, can you take a look?

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -63,12 +68,16 @@ def parse(String description) {
 	if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap.commandInt != 0x07 && descMap?.value) {
 		map = getBatteryPercentageResult(Integer.parseInt(descMap.value, 16))
-	} else if (descMap?.clusterInt == zigbee.ONOFF_CLUSTER && descMap.isClusterSpecific) {
+	}else if (isAduroSmartRemote()) {
```
nit: please put spaces after the `}`

adurosmart replied:
Fixed issues

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -63,12 +68,16 @@ def parse(String description) {
 		map = getBatteryPercentageResult(Integer.parseInt(descMap.value, 16))
-	} else if (descMap?.clusterInt == zigbee.ONOFF_CLUSTER && descMap.isClusterSpecific) {
+	}else if (isAduroSmartRemote()) {
+		def descMap = zigbee.parseDescriptionAsMap(description)
```
This line isn't needed because it's already done at the top of the function

adurosmart replied:
Fixed issues

tpmanley replied:
@adurosmart the line `def descMap = zigbee.parseDescriptionAsMap(description)` is still present but should be removed

adurosmart replied:
@tpmanley Thank you review,Fixed issues.

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +240,45 @@ private getModelBindings(model) {
+		    if (d0 == "00") {
+			buttonNumber = 2
+		    }
+		    if (d0 == "01") {
```
nit: this could be an `else if` instead

adurosmart replied:
Fixed issues

tpmanley replied:
@adurosmart this is still an `if` but should be changed to an `else if`

adurosmart replied:
Thank you review,Fixed issues.

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +240,45 @@ private getModelBindings(model) {
+		def list2 = descMap.data
+		buttonNumber = (list2[1] as int) + 1
+	}
+	if (buttonNumber !=0) {
```
nit: should have a space after the `!=`

adurosmart replied:
Fixed issues

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +240,45 @@ private getModelBindings(model) {
+		buttonNumber = (list2[1] as int) + 1
+	}
+	if (buttonNumber !=0) {
+		def childevent = createEvent(name: "button", value: "pushed",data: [buttonNumber: 1], isStateChange: true)
```
nit: should have a space before `data`

adurosmart replied:
Fixed issues

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -223,3 +240,45 @@ private getModelBindings(model) {
+	if (descMap.clusterInt == zigbee.ONOFF_CLUSTER) {
+		if (descMap.command == "01") {
+		    buttonNumber = 1
+		}else if (descMap.command == "00") {
```
nit: spaces before `else if` in this function too

adurosmart replied:
Fixed issues

tpmanley reviewed (dismissed):
Just a few nits but functionally it looks good

tpmanley commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -235,8 +235,15 @@ private getModelNumberOfButtons() {[
 	for(def endpoint : 1..modelNumberOfButtons[model]) {
 		bindings += zigbee.addBinding(zigbee.ONOFF_CLUSTER, ["destEndpoint" : endpoint])
+		if (isAduroSmartRemote()) {
+			if ((index==2) || (index==3)){
```
Do you mean to reference `endpoint` here? `index` is never incremented

adurosmart replied:
@tpmanley, Thank you review, i changed the index increment method.

tpmanley replied:
@adurosmart I think it would be better to avoid the introduction of an additional `index` variable and just use `endpoint`.  Alternatively, you could leave the `for` loop untouched and put the following after the `for` loop:

```
if (isAduroSmartRemote()) {
    bindings += zigbee.addBinding(zigbee.LEVEL_CONTROL_CLUSTER, ["destEndpoint" : 2]) + 
        zigbee.addBinding(zigbee.LEVEL_CONTROL_CLUSTER, ["destEndpoint" : 3])
}
```

shrakuma commented:
@adurosmart  can you please take a look at the above

tpmanley commented:
@shrakuma they were waiting on us to review the latest changes. I just missed that updates had been made.

@adurosmart the changes look good to me now. Unfortunately there is now a merge conflict due to other changes to the same file. Can you rebase your changes on the latest version and then ping me to do a final review and merge?

adurosmart commented:
@tpmanley  
Thank you review.
Is it using the 'git rebase master' command?
Thank you very much for your suggestions.

tpmanley commented:
@adurosmart yes it can be done with `rebase`, but you might find it easier to just manually make the changes to the latest version of the file. To do that, first make a copy of the existing file, then do this:

```
git fetch -p
git reset --hard upstream/master # This will remove your local changes so make sure you have a copy saved
<manually update the file with the changes from your saved copy>
<commit changes>
git push -f origin devws-for-most-target-limited-co_144
```

adurosmart committed:
Update zigbee-multi-button.groovy

@tpmanley
Thank you for your suggestions, (https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/13835#issuecomment-587027214), I followed your suggestions to update the code.
Thank you for reviewing the code again.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -66,6 +72,10 @@ def parseAttrMessage(description) {
+    	} else if (descMap?.clusterInt == zigbee.ONOFF_CLUSTER && descMap.isClusterSpecific) {
+		map = getButtonEvent(descMap)
 	} else if (descMap?.clusterInt == zigbee.ONOFF_CLUSTER && descMap.isClusterSpecific) {
 		map = getButtonEvent(descMap)
```
this code has been duplicated

adurosmart replied:
Fixed

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -233,13 +248,60 @@ private getSupportedButtonValues() {
+		sendEventToChild(buttonNumber, childevent)
+		def descriptionText = "$device.displayName button $buttonNumber was $buttonState"
+		return createEvent(name: "button", value: buttonState, data: [buttonNumber: buttonNumber], descriptionText: descriptionText, isStateChange: true)
+        } else {
```
nit: spacing

adurosmart replied:
Thank you review.
In these few lines of code, I didn't find any problems with word spacing. Maybe I didn't see it.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -233,13 +248,60 @@ private getSupportedButtonValues() {
+		    def d0 = data[0]
+		    if (d0 == "00") {
+			buttonNumber = 2
+		    }else if (d0 == "01") {
```
nit: spacing

adurosmart replied:
Thank you review.
Fixed.

greens commented on devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:
```diff
@@ -233,13 +248,60 @@ private getSupportedButtonValues() {
+		} else if (descMap.command == "00") {
+		    buttonNumber = 4
+		} else {
+		    return [:]
```
I don't think you need to short-circuit this function call here, you should be handling this correctly below.

adurosmart replied:
Thank you review.
Fixed.

adurosmart committed:
Fixed bugs and code style issues

@greens Thank you review.
Fixed bugs and code style issues.

Automated-House commented:
Hi @adurosmart I've paired a Eria dimming button switch V2.0 to SmartThings using this now published DTH. There's a couple problems:
1) Device isn't checking in to the hub at the `checkInterval` defined in the DTH, so it's being marked as offline in the SmartThings app
2) Battery isn't reporting correctly.
